vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java

```diff
@@ -18,31 +18,88 @@
 
 import io.vertx.mysqlclient.impl.datatype.DataFormat;
 import io.vertx.mysqlclient.impl.protocol.ColumnDefinition;
+import io.vertx.sqlclient.desc.ColumnDescriptor;
 import io.vertx.sqlclient.impl.RowDesc;
 
-import java.util.Arrays;
+import java.util.AbstractList;
 import java.util.Collections;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.List;
 
 public class MySQLRowDesc extends RowDesc {
 
-  private final ColumnDefinition[] columnDefinitions;
+  public final ColumnDefinition[] columnDefinitions;
   private final DataFormat dataFormat;
 
-  public MySQLRowDesc(ColumnDefinition[] columnDefinitions, DataFormat dataFormat) {
-    super(Collections.unmodifiableList(Stream.of(columnDefinitions)
-      .map(ColumnDefinition::name)
-      .collect(Collectors.toList())), Collections.unmodifiableList(Arrays.asList(columnDefinitions)));
+  private MySQLRowDesc(List<String> columnNames, List<ColumnDescriptor> columnDescriptors, ColumnDefinition[] columnDefinitions, DataFormat dataFormat) {
+    super(columnNames, columnDescriptors);
     this.columnDefinitions = columnDefinitions;
     this.dataFormat = dataFormat;
   }
 
-  public ColumnDefinition[] columnDefinitions() {
-    return columnDefinitions;
+
+  public static MySQLRowDesc create(ColumnDefinition[] columnDefinitions, DataFormat dataFormat) {
+    if (columnDefinitions.length == 0) {
+      return new MySQLRowDesc(Collections.emptyList(), Collections.emptyList(), columnDefinitions, dataFormat);
+    }
+    List<String> columnNames = new ColumnNames(columnDefinitions);
+    List<ColumnDescriptor> columnDescriptors = new ColumnDescriptors(columnDefinitions);
+    return new MySQLRowDesc(columnNames, columnDescriptors, columnDefinitions, dataFormat);
+  }
+
+  public int size() {
+    return columnDefinitions.length;
+  }
+
+  public ColumnDefinition get(int index) {
+    if (index < 0 || index >= size()) {
+      throw new IndexOutOfBoundsException();
+    }
+    return columnDefinitions[index];
   }
 
   public DataFormat dataFormat() {
     return dataFormat;
   }
+
+  private static class ColumnNames extends AbstractList<String> {
```

Comment:
Not sure what you mean, it extends `AbstractList` which throws `UnsupportedOperationException` on modifiers.

Reply:
If outside it's used as `ColumnNames` and never as `AbstractList` (if not for few/one single trivial method) you can save extending it.
It will make the JIT live a bit easier (not another list implementor bla bla bla) - but I'm ok whatever choice you made here, really

Reply:
Outside it's never used as `ColumnNames`, only as `List`:

https://github.com/eclipse-vertx/vertx-sql-client/blob/cf1da8a1652f5d1cefaec7c17cd892bfbb4d1448/vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java#L44-L46

https://github.com/eclipse-vertx/vertx-sql-client/blob/cf1da8a1652f5d1cefaec7c17cd892bfbb4d1448/vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java#L33-L34

https://github.com/eclipse-vertx/vertx-sql-client/blob/3b303d98f2410455eb47f01ac54d03f1c3dfccdf/vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowDesc.java#L62-L65